source/geod24/concurrency.d

```diff
@@ -1446,3 +1710,1545 @@ private:
+
+        return q_ent;
+    }
+
+    /***********************************************************************
+
+        Dequeue a `ChannelQueueEntry` from the given `entryq`
+
+        Walks the `entryq` until it finds a suitable entry or the queue
+        empties. If `dequeueEntry` is called from a select, it tries to
+        consume the `caller_sel` if the `peer_sel` is not currently consumed.
+        If it fails to consume the `caller_sel`, returns with a failure.
+
+        If selected queue entry is part of a select operation, it is also
+        consumed. If it consumes `caller_sel` but `peer_sel` was already
+        consumed whole select operation would fail and caller would need to try
+        again. This should be a rare case, where the `peer_sel` gets consumed by
+        someone else between the first if check which verifies that it is not
+        consumed and the point we actually try to consume it.
+
+        Param:
+            entryq = Queue to append the new ChannelQueueEntry
+            caller_sel = SelectState instace of the select call being executed
+            caller_sel_id = id of the select call being executed
+
+        Return:
+            a valid ChannelQueueEntry or null
+
+    ***********************************************************************/
+
+    private ChannelQueueEntry dequeueEntry (ref DList!ChannelQueueEntry entryq,
+                                            SelectState caller_sel = null, int caller_sel_id = 0) nothrow
+    {
+        while (!entryq.empty())
+        {
+            ChannelQueueEntry qent = entryq.front();
+            auto peer_sel = qent.select_state;
+
+            if ((!peer_sel || !peer_sel.isConsumed()) && qent.blocker.shouldBlock())
+            {
+                // If we are in a select call, try to consume the caller select
+                // if we can't consume the caller select, no need to continue
+                if (caller_sel && !caller_sel.tryConsume(caller_sel_id))
+                {
+                    return null;
+                }
+
+                // At this point, caller select is consumed.
+                // Try to consume the peer select if it exists
+                // If peer_sel was consumed by someone else, tough luck
+                // In that case, whole select will fail since we consumed the caller_sel
+                if ((!peer_sel || peer_sel.tryConsume(qent.sel_id)) && qent.blocker.stopTimer())
+                {
+                    entryq.removeFront();
+                    return qent;
+                }
+                else if (caller_sel)
+                {
+                    // Mark caller_sel failed
+                    caller_sel.id = -1;
+                    caller_sel.success = false;
+                    return null;
+                }
+            }
+
+            entryq.removeFront();
+        }
+
+        return null;
+    }
+
+private:
+    /// Internal data storage
+    FixedRingBuffer!T buffer;
```

Comment:
Do we really need to bring in a 315 LoC file just for this ? `FixedRingBuffer` seems rather generic, but we only iterate a single item at once and never resize the buffer. I only see 4 methods being used, `putBack`, `popFront`, `length`, `front`, and the initial capacity set.